## 1.	Evaluate text similarity of Amazon book search results by doing the following:
##### a.	Do a book search on Amazon. Manually copy the full book title (including subtitle) of each of the top 24 books listed in the first two pages of search results. 


In [1]:
import pandas as pd
import numpy as np
import requests
import json
import nltk
from bs4 import BeautifulSoup

In [2]:
best_books2019 = [
                "The Testaments: The Sequel to The Handmaid's Tale",
                "The Nickel Boys: A Novel",
                "Wild Game: My Mother, Her Lover, and Me",
                "Quichotte: A Novel",
                "The Starless Sea: A Novel",
                "Super Pumped: The Battle for Uber",
                "City of Girls: A Novel",
                "They Called Us Enemy",
                "The Silent Patient",
                "Maybe You Should Talk to Someone: A Therapist, HER Therapist, and Our Lives Revealed",
                "The Dutch House: A Novel",
                "Red at the Bone: A Novel",
                "Ask Again, Yes: A Novel",
                "The World That We Knew: A Novel",
                "On Earth We're Briefly Gorgeous: A Novel",
                "I Will Never See the World Again: The Memoir of an Imprisoned Writer",
                "The Water Dancer (Oprah's Book Club): A Novel",
                "Ninth House (Alex Stern)",
                "Three Women",
                "Things We Didn't Talk About When I Was a Girl: A Memoir",
                "Save Me the Plums: My Gourmet Memoir",
                "A Woman Is No Man: A Novel",
                "Once More We Saw Stars: A Memoir",
                "Mrs. Everything: A Novel"
               ]

best_books2019

["The Testaments: The Sequel to The Handmaid's Tale",
 'The Nickel Boys: A Novel',
 'Wild Game: My Mother, Her Lover, and Me',
 'Quichotte: A Novel',
 'The Starless Sea: A Novel',
 'Super Pumped: The Battle for Uber',
 'City of Girls: A Novel',
 'They Called Us Enemy',
 'The Silent Patient',
 'Maybe You Should Talk to Someone: A Therapist, HER Therapist, and Our Lives Revealed',
 'The Dutch House: A Novel',
 'Red at the Bone: A Novel',
 'Ask Again, Yes: A Novel',
 'The World That We Knew: A Novel',
 "On Earth We're Briefly Gorgeous: A Novel",
 'I Will Never See the World Again: The Memoir of an Imprisoned Writer',
 "The Water Dancer (Oprah's Book Club): A Novel",
 'Ninth House (Alex Stern)',
 'Three Women',
 "Things We Didn't Talk About When I Was a Girl: A Memoir",
 'Save Me the Plums: My Gourmet Memoir',
 'A Woman Is No Man: A Novel',
 'Once More We Saw Stars: A Memoir',
 'Mrs. Everything: A Novel']

#####  b) Compare each of the book titles based on cosine similarity, pairwise, to every other one.

In [4]:
from nltk.tokenize import word_tokenize
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.metrics.pairwise import cosine_similarity
from scipy import sparse

book_list_tokenized = []

for book in best_books2019:
    text = word_tokenize(book)
    book_list_tokenized.append(text)

In [6]:
# Vectorize the titles using CountVectorizer

count_vectorizer = CountVectorizer()
count_vectorizer.fit_transform(best_books2019)

<24x96 sparse matrix of type '<class 'numpy.int64'>'
	with 133 stored elements in Compressed Sparse Row format>

In [7]:
print(count_vectorizer.vocabulary_)

{'the': 76, 'testaments': 74, 'sequel': 64, 'to': 81, 'handmaid': 27, 'tale': 72, 'nickel': 44, 'boys': 10, 'novel': 47, 'wild': 88, 'game': 22, 'my': 42, 'mother': 40, 'her': 28, 'lover': 34, 'and': 4, 'me': 37, 'quichotte': 56, 'starless': 68, 'sea': 62, 'super': 71, 'pumped': 55, 'battle': 7, 'for': 21, 'uber': 82, 'city': 13, 'of': 48, 'girls': 24, 'they': 78, 'called': 12, 'us': 83, 'enemy': 19, 'silent': 66, 'patient': 53, 'maybe': 36, 'you': 95, 'should': 65, 'talk': 73, 'someone': 67, 'therapist': 77, 'our': 52, 'lives': 33, 'revealed': 59, 'dutch': 17, 'house': 29, 'red': 58, 'at': 6, 'bone': 8, 'ask': 5, 'again': 1, 'yes': 94, 'world': 92, 'that': 75, 'we': 86, 'knew': 32, 'on': 49, 'earth': 18, 're': 57, 'briefly': 11, 'gorgeous': 25, 'will': 89, 'never': 43, 'see': 63, 'memoir': 38, 'an': 3, 'imprisoned': 30, 'writer': 93, 'water': 85, 'dancer': 15, 'oprah': 51, 'book': 9, 'club': 14, 'ninth': 45, 'alex': 2, 'stern': 70, 'three': 80, 'women': 91, 'things': 79, 'didn': 16, '

In [40]:
tf_matrix = count_vectorizer.transform(best_books2019).toarray()
tfidfTran = TfidfTransformer(norm="l2")
tfidfTran.fit(tf_matrix)
tfidf_matrix = tfidfTran.transform(tf_matrix)
print(tfidf_matrix.shape)

(24, 96)


In [41]:
cos_similarity_matrix = (tfidf_matrix * tfidf_matrix.T).toarray()
print(cos_similarity_matrix)

[[1.         0.17529189 0.         0.         0.17529189 0.12013134
  0.         0.         0.18369366 0.07897009 0.18355948 0.14782248
  0.         0.1412254  0.         0.17513895 0.11768647 0.
  0.         0.         0.11884363 0.         0.         0.        ]
 [0.17529189 1.         0.         0.12697169 0.18762868 0.0673293
  0.08092806 0.         0.10295369 0.         0.19647814 0.15822602
  0.08092806 0.15116464 0.05839705 0.09815908 0.12596907 0.
  0.         0.         0.06660758 0.06827124 0.         0.09412714]
 [0.         0.         1.         0.         0.         0.
  0.         0.         0.         0.15601511 0.         0.
  0.         0.         0.         0.         0.         0.
  0.         0.         0.25343026 0.         0.         0.        ]
 [0.         0.12697169 0.         1.         0.12697169 0.
  0.11496001 0.         0.         0.         0.13296028 0.10707438
  0.11496001 0.10229583 0.08295424 0.         0.08524553 0.
  0.         0.         0.        

##### c) Which two titles are the most similar to each other? Which are the most dissimilar? Where do they rank, among the first 24 results?

In [42]:
length, width = cos_similarity_matrix.shape

In [43]:
print(cos_similarity_matrix[10,17])

0.26873929893960785


In [44]:
max_value = 0.0
row = None
col = None

for l in range(length):
    for w in range(width):
        if l == w:
            continue
        else:
            if cos_similarity_matrix[l,w] > max_value: 
                max_value = cos_similarity_matrix[l,w]
                row = l
                col = w
                
        
print("The most similar book titles have a cosine similarity of {}".format(max_value))
print("The {r} ranked booked is '{t}'".format(r=row, t=best_books2019[row]))
print("The {c} ranked booked is '{t}'".format(c=col, t=best_books2019[col]))

The most similar book titles have a cosine similarity of 0.26873929893960785
The 10 ranked booked is 'The Dutch House: A Novel'
The 17 ranked booked is 'Ninth House (Alex Stern)'


#### The most similiar books were Number 10'The Dutch House: A Novel' and 17 'Ninth House(Alex Stern)' with a cosine similarity of 29%

In [45]:
min_value = 1.0
row = None
col = None

for l in range(length):
    for w in range(width):
        if l == w:
            continue
        if cos_similarity_matrix[l,w] == 0:
            continue
        else:
            if cos_similarity_matrix[l,w] < min_value: 
                min_value = cos_similarity_matrix[l,w]
                row = l
                col = w
                
        
print("The most dissimilar book titles have a cosine similarity of {}".format(min_value))
print("The {r} ranked booked is '{t}'".format(r=row, t=best_books2019[row]))
print("The {c} ranked booked is '{t}'".format(c=col, t=best_books2019[col]))

The most dissimilar book titles have a cosine similarity of 0.03920627867468334
The 14 ranked booked is 'On Earth We're Briefly Gorgeous: A Novel'
The 16 ranked booked is 'The Water Dancer (Oprah's Book Club): A Novel'


#### The most dissimiliar books were Number 14'On Earth We're a Briefly Gorgeous: A Novel' and 16 'The Water Dance (Oprah's Book Club)' with a cosine similarity of 4%

## 2.	Now evaluate using a major search engine.
##### a.	Enter one of the book titles from question 1a into Google, Bing, or Yahoo!. Copy the capsule of the first organic result and the 20th organic result. Take web results only (i.e., not video results), and skip sponsored results. 


In [46]:
book_list_Google = [
    "The Testaments: The Sequel to The Handmaid's Tale", # Book 
    "The Testaments is a 2019 novel by Margaret Atwood. It is a sequel to The Handmaid's Tale (1985). The novel is set 15 years after the events of The Handmaid's Tale. It is narrated by Aunt Lydia, a character from the previous novel; Agnes, a young woman living in Gilead; and Daisy, a young woman living in Canada.", # 1st result from Google
    "Almost 35 years after “The Handmaid's Tale” shocked the world, Margaret Atwood has finally given birth to “The Testaments.”" #20th result from Google
]


##### b.	Run the same text similarity calculation that you used for question 1b on each of these capsules in comparison to the original query (book title). 

In [47]:
# Tokenize titles
book_list_Google_tokenized = []

for book in book_list_Google:
    text = word_tokenize(book)
    book_list_Google_tokenized.append(text)
    
# Vectorize the titles using CountVectorizer
count_vectorizer = CountVectorizer()
count_vectorizer.fit_transform(book_list_Google)

# Create tfidf matrix
tf_matrix = count_vectorizer.transform(book_list_Google).toarray()
tfidfTran = TfidfTransformer(norm="l2")
tfidfTran.fit(tf_matrix)
tfidf_matrix = tfidfTran.transform(tf_matrix)

# Pairwise
search_similarity_matrix = (tfidf_matrix * tfidf_matrix.T).toarray()
print(search_similarity_matrix)

[[1.         0.42947692 0.52043953]
 [0.42947692 1.         0.3008865 ]
 [0.52043953 0.3008865  1.        ]]


In [48]:
search_similarity_matrix[0,1]

0.42947692329083703

In [49]:
search_similarity_matrix[0,2]

0.520439527276546

##### The 1st google query had a similiarity around 43% and the 20th around 52%

##### c.	Which one has the highest similarity measure? 

## 3.	Take the dataset from Homework 4 and run it through an RNN, an LSTM, and a Bidirectional RNN.
###### a.	Report your accuracy, precision, and F1 for each class (‘O’ counts as a class, ‘PAD’ should be ignored).
###### b.  Turn in a Confusion Matrix of your results for each method.

In [50]:
import re

f = open('train.txt','r')
tagged_sentence = []
sentence = []

for line in f:
    if len(line)==0 or line.startswith('-DOCSTART') or line[0]=='\n':
        if len(sentence) >0:
            tagged_sentence.append(sentence)
            sentence = []
        continue
    splits = line.split(' ')
    splits[-1] = re.sub(r'\n','',splits[-1])
    word = splits[0].lower()
    sentence.append([word,splits[-1]])

In [51]:
sentence_list = []
tag_list = []

for tagged_sentence in tagged_sentence:
    sentence, tag_info = zip(*tagged_sentence)
    sentence_list.append(list(sentence))
    tag_list.append(list(tag_info))

In [52]:
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.utils import to_categorical
from sklearn.model_selection import train_test_split
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Embedding, LSTM, Bidirectional, TimeDistributed, SimpleRNN
from tensorflow.keras.optimizers import Adam
import tensorflow as tf

max_words = 4000

vocab = Tokenizer(num_words=max_words, oov_token='OOV')
vocab.fit_on_texts(sentence_list)

tag = Tokenizer()
tag.fit_on_texts(tag_list)

In [53]:
vocab_size = max_words
tag_size = len(tag.word_index) + 1
print(vocab_size, tag_size)

4000 10


In [54]:
X_train = vocab.texts_to_sequences(sentence_list)
y_train = tag.texts_to_sequences(tag_list)

In [55]:
index2word = vocab.index_word
index2ner = tag.index_word
index2ner[0] = 'PAD'

In [56]:
indextarget = dict([(value, key) for (key, value) in index2ner.items()])

In [57]:
indextarget

{'o': 1,
 'b-loc': 2,
 'b-per': 3,
 'b-org': 4,
 'i-per': 5,
 'i-org': 6,
 'b-misc': 7,
 'i-loc': 8,
 'i-misc': 9,
 'PAD': 0}

In [58]:
ex_decoded = []
for index in X_train[0] : 
    ex_decoded.append(index2word[index])

In [59]:
max_len = 70
X_train = pad_sequences(X_train, maxlen=max_len)
y_train = pad_sequences(y_train, maxlen=max_len)

In [60]:
X_train, X_test, y_train, y_test = train_test_split(X_train, y_train, test_size=.2, random_state=42)

In [61]:
y_train = to_categorical(y_train, num_classes=tag_size)
y_test = to_categorical(y_test, num_classes=tag_size)

## Simple RNN Model

In [62]:
model = Sequential()
model.add(Embedding(input_dim=vocab_size, output_dim=128, input_length=max_len, mask_zero=True))
model.add(SimpleRNN(units=500, return_sequences=True, recurrent_dropout=0.2, dropout=0.2))
model.add(TimeDistributed(Dense(tag_size, activation='softmax')))
model.compile(loss='categorical_crossentropy', optimizer="adam", metrics=['accuracy', tf.keras.metrics.Precision(),tf.keras.metrics.Recall()] )
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_1 (Embedding)      (None, 70, 128)           512000    
_________________________________________________________________
simple_rnn_1 (SimpleRNN)     (None, 70, 500)           314500    
_________________________________________________________________
time_distributed_1 (TimeDist (None, 70, 10)            5010      
Total params: 831,510
Trainable params: 831,510
Non-trainable params: 0
_________________________________________________________________


In [63]:
history = model.fit(X_train, y_train, batch_size=128, epochs=4,  validation_data=(X_test, y_test))

Train on 11232 samples, validate on 2809 samples
Epoch 1/4
11232/11232 [==============================] - 60s 5ms/sample - loss: 0.1747 - acc: 0.8169 - precision_1: 0.8670 - recall_1: 0.7277 - val_loss: 0.1178 - val_acc: 0.8580 - val_precision_1: 0.8941 - val_recall_1: 0.8200
Epoch 2/4
11232/11232 [==============================] - 57s 5ms/sample - loss: 0.0877 - acc: 0.8772 - precision_1: 0.9346 - recall_1: 0.8292 - val_loss: 0.0666 - val_acc: 0.9096 - val_precision_1: 0.9557 - val_recall_1: 0.8625
Epoch 3/4
11232/11232 [==============================] - 61s 5ms/sample - loss: 0.0602 - acc: 0.9168 - precision_1: 0.9550 - recall_1: 0.8761 - val_loss: 0.0518 - val_acc: 0.9298 - val_precision_1: 0.9621 - val_recall_1: 0.9018
Epoch 4/4
11232/11232 [==============================] - 59s 5ms/sample - loss: 0.0486 - acc: 0.9313 - precision_1: 0.9572 - recall_1: 0.9088 - val_loss: 0.0458 - val_acc: 0.9356 - val_precision_1: 0.9558 - val_recall_1: 0.9220


In [64]:
SRNN_metrics = model.evaluate(X_test, y_test, verbose=1)
print("Simple RNN Metrics(Test loss, Test Accuracy, Precision, & Recall): ")
print(SRNN_metrics)

2809/2809 [==============================] - 9s 3ms/sample - loss: 0.0458 - acc: 0.9356 - precision_1: 0.9558 - recall_1: 0.9220
Simple RNN Metrics(Test loss, Test Accuracy, Precision, & Recall): 
[0.0458189492992826, 0.9356094, 0.95580363, 0.92200196]


In [65]:
import numpy as np
from sklearn.metrics import f1_score
from sklearn.metrics import precision_score
y_pred = model.predict(np.array(X_test))
y_pred = np.argmax(y_pred,axis=-1)
true = np.argmax(np.array(y_test),-1)
true

array([[0, 0, 0, ..., 1, 1, 1],
       [0, 0, 0, ..., 1, 1, 1],
       [0, 0, 0, ..., 1, 1, 1],
       ...,
       [0, 0, 0, ..., 1, 1, 1],
       [0, 0, 0, ..., 2, 8, 1],
       [0, 0, 0, ..., 3, 1, 1]], dtype=int64)

In [66]:
count = 0
total = 0
pred = []
actual = []

for prediction in range(y_pred.shape[0]):                       
    for i in range(max_len):                            
        if true[prediction][i]==0:  #skip PAD value so not in DF
            pass
        else:                                                 
            if true[prediction][i]==y_pred[prediction][i]:      
                count = count+1                                
                total=total+1                                      
            pred.append(y_pred[prediction][i])                   
            actual.append(true[prediction][i])

In [67]:
from sklearn.metrics import precision_recall_fscore_support
import pandas as pd
simpRNNDF = pd.DataFrame(precision_recall_fscore_support(y_true=actual, y_pred=pred, average=None))

In [68]:
simpRNNDF.index = ["Precision", "Recall", "FScore" ,"Support"]
simpRNNDF.columns = [index2ner[col] for col in list(set(actual))]
simpRNNDF

o        b-loc        b-per        b-org       i-per  \
Precision      0.948596     0.874613     0.895769     0.862022    0.856957   
Recall         0.991879     0.794097     0.692185     0.520627    0.778040   
FScore         0.969755     0.832413     0.780927     0.649177    0.815594   
Support    34233.000000  1423.000000  1254.000000  1212.000000  847.000000   

                i-org      b-misc       i-loc      i-misc  
Precision    0.774481    0.778878    0.708738    0.694805  
Recall       0.359010    0.699259    0.561538    0.467249  
FScore       0.490602    0.736924    0.626609    0.558747  
Support    727.000000  675.000000  260.000000  229.000000

In [69]:
from sklearn.metrics import confusion_matrix
# Compute confusion matrix
simpRNNMatrix = confusion_matrix(actual, pred)

simpRNNMatrix

array([[33955,    32,    46,    18,    69,    29,    47,    10,    27],
       [  214,  1130,     5,    48,     0,     3,    20,     2,     1],
       [  346,     4,   868,    19,    13,     2,     1,     1,     0],
       [  417,    78,    29,   631,     1,     5,    45,     2,     4],
       [  163,     0,    13,     0,   659,    10,     1,     1,     0],
       [  369,    17,     6,     5,    16,   261,     6,    39,     8],
       [  166,    18,     2,     8,     0,     3,   472,     1,     5],
       [   74,     7,     0,     3,    10,    15,     3,   146,     2],
       [   91,     6,     0,     0,     1,     9,    11,     4,   107]],
      dtype=int64)

## LSTM Model

In [70]:
model = Sequential()
model.add(Embedding(input_dim=vocab_size, output_dim=128, input_length=max_len, mask_zero=True))
model.add(LSTM(units=500, return_sequences=True, recurrent_dropout=0.2, dropout=0.2))
model.add(TimeDistributed(Dense(tag_size, activation='softmax')))
model.compile(loss='categorical_crossentropy', optimizer="adam", metrics=['accuracy', tf.keras.metrics.Precision(),tf.keras.metrics.Recall()])
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_2 (Embedding)      (None, 70, 128)           512000    
_________________________________________________________________
lstm (LSTM)                  (None, 70, 500)           1258000   
_________________________________________________________________
time_distributed_2 (TimeDist (None, 70, 10)            5010      
Total params: 1,775,010
Trainable params: 1,775,010
Non-trainable params: 0
_________________________________________________________________


In [71]:
history = model.fit(X_train, y_train, batch_size=128, epochs=4,  validation_data=(X_test, y_test))

Train on 11232 samples, validate on 2809 samples
Epoch 1/4
11232/11232 [==============================] - 295s 26ms/sample - loss: 0.1914 - acc: 0.8235 - precision_2: 0.8630 - recall_2: 0.6827 - val_loss: 0.1367 - val_acc: 0.8366 - val_precision_2: 0.8700 - val_recall_2: 0.8102
Epoch 2/4
11232/11232 [==============================] - 296s 26ms/sample - loss: 0.1054 - acc: 0.8457 - precision_2: 0.9210 - recall_2: 0.8116 - val_loss: 0.0829 - val_acc: 0.8711 - val_precision_2: 0.9578 - val_recall_2: 0.8256
Epoch 3/4
11232/11232 [==============================] - 278s 25ms/sample - loss: 0.0763 - acc: 0.8837 - precision_2: 0.9561 - recall_2: 0.8339 - val_loss: 0.0691 - val_acc: 0.9007 - val_precision_2: 0.9554 - val_recall_2: 0.8578
Epoch 4/4
11232/11232 [==============================] - 291s 26ms/sample - loss: 0.0632 - acc: 0.9087 - precision_2: 0.9591 - recall_2: 0.8630 - val_loss: 0.0589 - val_acc: 0.9186 - val_precision_2: 0.9593 - val_recall_2: 0.8779


In [72]:
lstm_metrics = model.evaluate(X_test, y_test, verbose=1)
print("LSTM Metrics(Test loss Test Accuracy, Precision, & Recall): ")
print(lstm_metrics)

2809/2809 [==============================] - 30s 11ms/sample - loss: 0.0589 - acc: 0.9186 - precision_2: 0.9593 - recall_2: 0.8779
LSTM Metrics(Test loss Test Accuracy, Precision, & Recall): 
[0.05887940408374839, 0.9186246, 0.95929503, 0.8778512]


In [73]:
import numpy as np
from sklearn.metrics import f1_score
from sklearn.metrics import precision_score
y_pred = model.predict(np.array(X_test))
y_pred = np.argmax(y_pred,axis=-1)
true = np.argmax(np.array(y_test),-1)
true

array([[0, 0, 0, ..., 1, 1, 1],
       [0, 0, 0, ..., 1, 1, 1],
       [0, 0, 0, ..., 1, 1, 1],
       ...,
       [0, 0, 0, ..., 1, 1, 1],
       [0, 0, 0, ..., 2, 8, 1],
       [0, 0, 0, ..., 3, 1, 1]], dtype=int64)

In [74]:
count = 0
total = 0
pred = []
actual = []

for prediction in range(y_pred.shape[0]):                       
    for i in range(max_len):                            
        if true[prediction][i]==0:  #skip PAD value so not in DF
            pass
        else:                                                 
            if true[prediction][i]==y_pred[prediction][i]:      
                count = count+1                                
                total=total+1                                      
            pred.append(y_pred[prediction][i])                   
            actual.append(true[prediction][i])

In [75]:
from sklearn.metrics import precision_recall_fscore_support
import pandas as pd
lstmDF = pd.DataFrame(precision_recall_fscore_support(y_true=actual, y_pred=pred, average=None))


C:\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64\lib\site-packages\sklearn\metrics\_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [76]:
lstmDF = lstmDF.drop(columns = [0])

In [77]:
lstmDF.index = ["Precision", "Recall", "FScore" ,"Support"]
lstmDF.columns = [index2ner[col] for col in list(set(actual))]
lstmDF

o        b-loc        b-per        b-org       i-per  \
Precision      0.946112     0.772535     0.686459     0.795756    0.731757   
Recall         0.988812     0.770907     0.755981     0.247525    0.840614   
FScore         0.966991     0.771720     0.719545     0.377596    0.782418   
Support    34233.000000  1423.000000  1254.000000  1212.000000  847.000000   

                i-org      b-misc       i-loc      i-misc  
Precision    0.600962    0.752049    0.584270    0.640000  
Recall       0.171939    0.543704    0.400000    0.139738  
FScore       0.267380    0.631126    0.474886    0.229391  
Support    727.000000  675.000000  260.000000  229.000000

In [78]:
from sklearn.metrics import confusion_matrix
# Compute confusion matrix
lstmMatrix = confusion_matrix(actual, pred)

lstmMatrix

array([[    0,     0,     0,     0,     0,     0,     0,     0,     0,
            0],
       [    2, 33850,    14,   133,    35,   133,    17,    38,     5,
            6],
       [    0,   227,  1097,    42,    20,     4,     4,    27,     2,
            0],
       [    0,   264,     7,   948,     9,    25,     1,     0,     0,
            0],
       [    0,   509,   134,   224,   300,     5,     8,    31,     1,
            0],
       [    0,   116,     2,    16,     0,   712,     0,     1,     0,
            0],
       [    0,   418,    44,     7,     3,    73,   125,    14,    41,
            2],
       [    1,   175,    80,     5,    10,     3,    13,   367,    11,
           10],
       [    2,    85,    30,     6,     0,    14,    15,     4,   104,
            0],
       [    2,   134,    12,     0,     0,     4,    25,     6,    14,
           32]], dtype=int64)

## Bidirectional Model

In [79]:
model = Sequential()
model.add(Embedding(input_dim=vocab_size, output_dim=128, input_length=max_len, mask_zero=True))
model.add(Bidirectional(SimpleRNN(units=500, return_sequences=True, recurrent_dropout=0.2, dropout=0.2)))
model.add(TimeDistributed(Dense(tag_size, activation='softmax')))
model.compile(loss='categorical_crossentropy', optimizer="adam", metrics=['accuracy', tf.keras.metrics.Precision(),tf.keras.metrics.Recall()])
model.summary()

Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_3 (Embedding)      (None, 70, 128)           512000    
_________________________________________________________________
bidirectional (Bidirectional (None, 70, 1000)          629000    
_________________________________________________________________
time_distributed_3 (TimeDist (None, 70, 10)            10010     
Total params: 1,151,010
Trainable params: 1,151,010
Non-trainable params: 0
_________________________________________________________________


In [80]:
history = model.fit(X_train, y_train, batch_size=128, epochs=4,  validation_data=(X_test, y_test))

Train on 11232 samples, validate on 2809 samples
Epoch 1/4
11232/11232 [==============================] - 114s 10ms/sample - loss: 0.1579 - acc: 0.8283 - precision_3: 0.8746 - recall_3: 0.7635 - val_loss: 0.0910 - val_acc: 0.8779 - val_precision_3: 0.9240 - val_recall_3: 0.8456
Epoch 2/4
11232/11232 [==============================] - 116s 10ms/sample - loss: 0.0734 - acc: 0.8959 - precision_3: 0.9452 - recall_3: 0.8573 - val_loss: 0.0562 - val_acc: 0.9246 - val_precision_3: 0.9600 - val_recall_3: 0.8877
Epoch 3/4
11232/11232 [==============================] - 115s 10ms/sample - loss: 0.0486 - acc: 0.9335 - precision_3: 0.9594 - recall_3: 0.9074 - val_loss: 0.0419 - val_acc: 0.9448 - val_precision_3: 0.9617 - val_recall_3: 0.9273
Epoch 4/4
11232/11232 [==============================] - 175s 16ms/sample - loss: 0.0372 - acc: 0.9498 - precision_3: 0.9645 - recall_3: 0.9349 - val_loss: 0.0361 - val_acc: 0.9524 - val_precision_3: 0.9623 - val_recall_3: 0.9432


In [81]:
Bi_metrics = model.evaluate(X_test, y_test, verbose=1)
print("Bidirectional Metrics(Test loss Test Accuracy, Precision, & Recall): ")
print(Bi_metrics)

2809/2809 [==============================] - 24s 9ms/sample - loss: 0.0361 - acc: 0.9524 - precision_3: 0.9623 - recall_3: 0.9432
Bidirectional Metrics(Test loss Test Accuracy, Precision, & Recall): 
[0.03614134113703834, 0.952374, 0.96229714, 0.94322073]


In [82]:
from sklearn.metrics import f1_score
from sklearn.metrics import precision_score
y_pred = model.predict(np.array(X_test))
y_pred = np.argmax(y_pred,axis=-1)
true = np.argmax(np.array(y_test),-1)

In [83]:
count = 0
total = 0
pred = []
actual = []

for prediction in range(y_pred.shape[0]):                       
    for i in range(max_len):                            
        if true[prediction][i]==0:  #skip PAD value so not in DF
            pass
        else:                                                 
            if true[prediction][i]==y_pred[prediction][i]:      
                count = count+1                                
                total=total+1                                      
            pred.append(y_pred[prediction][i])                   
            actual.append(true[prediction][i])                

In [84]:
from sklearn.metrics import precision_recall_fscore_support
import pandas as pd
birdirectionalDF = pd.DataFrame(precision_recall_fscore_support(y_true=actual, y_pred=pred, average=None))

In [85]:
birdirectionalDF.index = ["Precision", "Recall", "FScore" ,"Support"]
birdirectionalDF.columns = [index2ner[col] for col in list(set(actual))]
birdirectionalDF

o        b-loc        b-per        b-org       i-per  \
Precision      0.966453     0.913686     0.834504     0.848309    0.879093   
Recall         0.991353     0.825720     0.852472     0.724422    0.824085   
FScore         0.978745     0.867479     0.843393     0.781486    0.850701   
Support    34233.000000  1423.000000  1254.000000  1212.000000  847.000000   

                i-org      b-misc       i-loc      i-misc  
Precision    0.806452    0.881533    0.909677    0.887097  
Recall       0.550206    0.749630    0.542308    0.480349  
FScore       0.654129    0.810248    0.679518    0.623229  
Support    727.000000  675.000000  260.000000  229.000000

In [86]:
from sklearn.metrics import confusion_matrix
# Compute confusion matrix
confMatrix = confusion_matrix(actual, pred)

confMatrix


array([[33937,    18,    98,    55,    50,    49,    13,     3,    10],
       [  163,  1175,    20,    44,     3,     5,    10,     3,     0],
       [  148,     7,  1069,    21,     9,     0,     0,     0,     0],
       [  208,    40,    43,   878,     5,    10,    28,     0,     0],
       [  121,     0,    22,     0,   698,     3,     1,     2,     0],
       [  248,    18,    12,    25,    16,   400,     3,     5,     0],
       [  136,    10,     9,     9,     1,     1,   506,     1,     2],
       [   70,    12,     4,     3,     7,    20,     1,   141,     2],
       [   84,     6,     4,     0,     5,     8,    12,     0,   110]],
      dtype=int64)

##### The Bidirectional RNN had the best overall test accuracy at 96% compared to simpleRNN(93.5%) & lstm(91.8%).  o class had the highest numbers(precision, recall & f1 score) for all 3 tests. 